enso-org/enso
# Search bar for settings

### Pull Request Description
- Close https://github.com/enso-org/cloud-v2/issues/1198
  - Add search bar for settings
    - Note that, *unlike* the asset search bar, it has neither suggestions nor tags (name:, description:) - I don't think those add any value here (and definitely would complicate the implementation by a lot)
  - Convert settings objects to JS so that they can be filtered by search

UX changes required for this PR:
- Switched `MembersSettingsSection` to use `react-query` to avoid loading the members list every time a user types into the search bar

Incidental changes added to make this PR simpler:
- Refactor search bar to be provided by a React Context so that state no longer needs to be hoisted all the way up into `Dashboard` and back down to `TopBar`

### Important Notes
- While I think highlighting the tab of the currently visible section would be a nice UX upgrade, I think it might be a bit complex for at least this initial PR.
  - It's definitely possible, but almost certainly requires Intersection Observers, plus a new constant to map each Settings Section ID back to its parent Settings Tab ID

### Screencast

Note: The end of this recording contains a bug where the "user account" section was not being matched. This bug has since been fixed.

https://github.com/enso-org/enso/assets/4046547/1023dbd6-907d-4add-a41a-d4b59e46f9b2

### Code Review tips
- `settingsData.ts` contains the settings converted to JS objects
- `Settings.tsx` contains the new filtering logic
- `Settings/Settings{Tab,Section,Entry,InputEntry,CustomEntry}` contains the components responsible for rendering the new JS objects
  - The old components which manually constructed the tabs from their constituent components have been removed

### Testing instructions
- Mostly testing that the search works as expected
- Partly testing to make sure nothing was broken by the refactors - either CSS wise or otherwise.
  - Most things *should* retain 100% of their original logic - search for `SettingsEntryType.custom` in `settingsData`

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
